cmd/checkpoint/main.go

```diff
@@ -0,0 +1,268 @@
+package main
+
+import (
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"net/http"
+	"os"
+	"path/filepath"
+	"strings"
+	"time"
+
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/api/v1"
+	clientset "k8s.io/kubernetes/pkg/client/clientset_generated/release_1_2"
+	"k8s.io/kubernetes/pkg/client/restclient"
+	"k8s.io/kubernetes/pkg/client/unversioned/clientcmd"
+)
+
+var manifestBytes = []byte(`
+{
+"apiVersion": "v1",
+"kind": "Pod",
+"metadata": {
+  "name": "temp-apiserver",
+  "namespace": "kube-system"
+},
+"spec": {}
+}
+`)
+
+const (
+	kubeletAPIPodsURL = "http://localhost:10255/pods"
+	ignorePath        = "/srv/kubernetes/manifests"
+	activePath        = "/etc/kubernetes/manifests"
+	manifestFilename  = "apiserver.json"
+	kubeconfigPath    = "/etc/kubernetes/kubeconfig"
+	tlsPath           = "/etc/kubernetes/tls"
+)
+
+var (
+	tempAPIServer = []byte("temp-apiserver")
+	kubeAPIServer = []byte("kube-apiserver")
+	activeFile    = filepath.Join(activePath, manifestFilename)
+	ignoreFile    = filepath.Join(ignorePath, manifestFilename)
+	secureAPIAddr = fmt.Sprintf("https://%s:%s", os.Getenv("KUBERNETES_SERVICE_HOST"), os.Getenv("KUBERNETES_SERVICE_PORT_HTTPS"))
+)
+
+func main() {
+	log.Println("begin apiserver checkpointing...")
+	run()
+}
+
+func run() {
+	var tempAPIServerManifest v1.Pod
+	if err := json.Unmarshal(manifestBytes, &tempAPIServerManifest); err != nil {
+		log.Fatal(err)
+	}
+	client := newAPIClient()
+	for {
+		var podList v1.PodList
+		if err := json.Unmarshal(getPodsFromKubeletAPI(), &podList); err != nil {
+			log.Fatal(err)
+		}
+		switch {
+		case bothAPIServersRunning(podList):
+			log.Println("both temp and kube apiserver running, removing temp apiserver")
+			// Both the self-hosted API Server and the temp API Server are running.
+			// Remove the temp API Server manifest from the config dir so that the
+			// kubelet will stop it.
+			if err := os.Remove(activeFile); err != nil {
+				log.Println(err)
+			}
+		case kubeSystemAPIServerRunning(podList):
+			log.Println("kube-apiserver found, creating temp-apiserver manifest")
+			// The self-hosted API Server is running. Let's snapshot the pod,
+			// clean it up a bit, and then save it to the ignore path for
+			// later use.
+			tempAPIServerManifest.Spec = parseAPIPodSpec(podList)
+			convertSecretsToVolumeMounts(client, &tempAPIServerManifest.Spec)
+			writeManifest(tempAPIServerManifest)
+			log.Printf("finished creating temp-apiserver manifest at %s\n", ignoreFile)
+
+		default:
+			log.Println("no apiserver running, installing temp apiserver static manifest")
+			b, err := ioutil.ReadFile(ignoreFile)
```

Comment:
nit: s/ingoreFile/checkpointManifest -- or something a little more descriptive. So maybe the pair is something like:
`checkpointManifest`
`activeManifest`